File: src/io/tasks/to_csv.cc
/* Copyright 2021 NVIDIA Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include <fstream>

#include "io/tasks/to_csv.h"
#include "io/file_util.h"
#include "column/detail/column.h"
#include "string/converter.h"
#include "util/type_dispatch.h"

namespace legate {
namespace pandas {
namespace io {
namespace csv {

using namespace Legion;

using ToCSVArgs = ToCSVTask::ToCSVArgs;

void ToCSVTask::ToCSVArgs::cleanup(void)
{
  for (auto &column : columns) column.destroy();
}

using ColumnView = detail::Column;

namespace detail {

template <bool nullable>
struct Print {
  template <TypeCode CODE,
            std::enable_if_t<is_numeric_type<CODE>::value || CODE == TypeCode::STRING> * = nullptr>
  void operator()(std::ostream &stream,
                  const ColumnView &column,
                  size_t idx,
                  const std::string &na_rep)
  {
    using VAL = pandas_type_of<CODE>;
    string::detail::Converter<VAL> convert;
    if (nullable)
      stream << (column.bitmask().get(idx) ? convert(column.element<VAL>(idx)) : na_rep);
    else
      stream << convert(column.element<VAL>(idx));
  }

  template <TypeCode CODE,
            std::enable_if_t<!is_numeric_type<CODE>::value && CODE != TypeCode::STRING> * = nullptr>
  void operator()(std::ostream &stream,
                  const ColumnView &column,
                  size_t idx,
                  const std::string &na_rep)
  {
    assert(false);
  }
};

void print_column(std::ostream &stream,
                  const ColumnView &column,
                  size_t idx,
                  const std::string &na_rep)
{
  if (column.nullable())
    type_dispatch(column.code(), Print<true>{}, stream, column, idx, na_rep);
  else
    type_dispatch(column.code(), Print<false>{}, stream, column, idx, na_rep);
}

void print_column(std::ostream &stream,
                  const std::string &column,
                  size_t idx,
                  const std::string &na_rep)
{
  stream << column;
}

template <typename Column>
void print_line(std::ostream &stream,
                const std::vector<Column> &columns,
                size_t row_idx,
                const ToCSVArgs &args)
{
  auto num_columns = columns.size();
  for (auto col_idx = 0; col_idx < num_columns; ++col_idx) {
    print_column(stream, columns[col_idx], row_idx, args.na_rep);
    stream << (col_idx == num_columns - 1 ? args.line_terminator : args.sep);
  }
}

}  // namespace detail

/*static*/ void ToCSVTask::cpu_variant(const Task *task,
                                       const std::vector<PhysicalRegion> &regions,
                                       Context context,
                                       Runtime *runtime)
{
  Deserializer ctx{task, regions};

  ToCSVArgs args;
  deserialize(ctx, args);

  std::vector<ColumnView> columns;
  for (auto &column : args.columns) columns.push_back(column.view());

  auto task_id = static_cast<uint32_t>(task->index_point[0]);
  auto path    = args.partition
                ? get_partition_filename(std::move(args.path), ".", "", args.num_pieces, task_id)
                : args.path;

  std::ofstream ofs;
  ofs.open(path, std::ofstream::out | std::ofstream::trunc);

  auto num_columns = args.column_names.size();
  if (args.header) detail::print_line(ofs, args.column_names, 0, args);

  std::stringstream ss{};
  auto size = columns.front().size();
  for (auto row_idx = 0; row_idx < size; ++row_idx) {
    detail::print_line(ss, columns, row_idx, args);
    if ((row_idx + 1) % args.chunksize == 0 || size - 1 == row_idx) {
      ofs << ss.str();
      ss = std::stringstream{};
    }
  }

  ofs.close();
}

void deserialize(Deserializer &ctx, ToCSVArgs &args)
{
  deserialize(ctx, args.num_pieces);
  deserialize(ctx, args.chunksize);
  deserialize(ctx, args.partition);
  deserialize(ctx, args.header);

  deserialize(ctx, args.path);
  deserialize(ctx, args.sep);
  deserialize(ctx, args.na_rep);
  deserialize(ctx, args.line_terminator);

  uint32_t num_columns = 0;
  deserialize(ctx, num_columns);
  args.column_names.resize(num_columns);
  args.columns.resize(num_columns);
  deserialize(ctx, args.column_names, false);
  deserialize(ctx, args.columns, false);
}

static void __attribute__((constructor)) register_tasks(void) { ToCSVTask::register_variants(); }

}  // namespace csv
}  // namespace io
}  // namespace pandas
}  // namespace legate

File: tests/io/to_csv.py
# Copyright 2021 NVIDIA Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import glob
import itertools
import os
import shutil
import tempfile

import pandas as pd
from numpy.random import permutation

from legate import pandas as lp

n = 100

data = {
    "col_a": range(n, 0, -1),
    "col_b": range(n),
    "col_c": ["prefix" + str(i) * 3 for i in range(n)],
}

dtype = {"col_a": "int32", "col_b": "float64", "col_c": "string"}

column_names = list(data.keys())

all_params = {
    "params_header": [True, False],
    "params_index": [True, False],
    "params_partition": [True, False],
    "params_columns": [None, [column_names[0], column_names[2]]],
    "params_sep": [",", "|"],
}


def sanitize_column_names(names):
    return column_names if names is None else names


def read_csv(fn, path, conf, index):
    sep = conf["sep"]

    if conf["index"]:
        if conf["header"]:
            return fn(
                path,
                sep=sep,
                dtype=dtype,
                index_col=list(range(index.nlevels)),
            )
        else:
            index_names = [f"ind_{lvl}" for lvl in range(index.nlevels)]
            names = index_names + sanitize_column_names(conf["columns"])
            for index_name in index_names:
                dtype[index_name] = "int64"
            return fn(
                path,
                sep=sep,
                names=names,
                dtype=dtype,
                index_col=names[: index.nlevels],
            )
    else:
        if conf["header"]:
            return fn(path, sep=sep, dtype=dtype)
        else:
            names = sanitize_column_names(conf["columns"])
            return fn(path, sep=sep, names=names, dtype=dtype)


for index in [
    pd.Index(permutation(n)),
    pd.MultiIndex.from_arrays(
        [permutation(n), permutation(n)], names=("icol1", "icol2")
    ),
    pd.RangeIndex(3 * n + 1, 1, -3, name="k"),
]:
    df = pd.DataFrame(data, index=index)
    for column in column_names:
        df[column] = df[column].astype(dtype[column])

    ldf = lp.DataFrame(df)

    for tpl in itertools.product(*all_params.values()):
        conf = {
            "header": tpl[0],
            "index": tpl[1],
            "partition": tpl[2],
            "columns": tpl[3],
            "sep": tpl[4],
        }

        print(f"Index type: {type(index)}, Conf: {conf}")

        path = os.path.join(os.path.dirname(__file__), "files")
        tmp_dir = tempfile.mkdtemp(dir=path)
        out_path = os.path.sep.join([tmp_dir, "out.csv"])

        print(f"Dump to {out_path}")

        try:
            ldf.to_csv(out_path, **conf)
            if conf["partition"]:
                partitions = sorted(glob.glob(out_path + ".*"))
                dfs = [
                    read_csv(pd.read_csv, part, conf, index)
                    for part in partitions
                ]
                df_copy = pd.concat(dfs)
            else:
                df_copy = read_csv(pd.read_csv, out_path, conf, index)

            columns = sanitize_column_names(conf["columns"])
            if conf["index"]:
                assert ldf[columns].equals(df_copy)
            else:
                assert (
                    ldf[columns]
                    .reset_index(drop=True)
                    .equals(df_copy.reset_index(drop=True))
                )
        finally:
            shutil.rmtree(tmp_dir)
